### ISSUE TITLE
Resuming training restarts at much higher loss

### ISSUE THREAD

**Issue by conceptofmind** [UKPLab/sentence-transformers#3756]
Hi Tom,

I am currently training models using this script:
```python
import os
os.environ["WANDB_MODE"] = "offline"
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True,max_split_size_mb:128"
os.environ["HF_HOME"] = "/e/project1/reformo/enrico/.hf_cache"

import torch
from torch import Tensor
from sentence_transformers.sentence_transformer.evaluation import InformationRetrievalEvaluator
from sentence_transformers.sentence_transformer.losses import CachedMultipleNegativesRankingLoss, MatryoshkaLoss
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerModelCardData,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments
)
from sentence_transformers.sentence_transformer.training_args import BatchSamplers
from datasets import Dataset, load_dataset
from collections.abc import Iterable

def prepare_test(test_set: str):
    ds = load_dataset("parquet", data_dir=test_set, split="train")

    queries = {}
    corpus = {}
    relevant_docs = {}

    for i, row in enumerate(ds):
        qid = f"q{i}"
        did = f"d{i}"

        queries[qid] = row["question"]
        corpus[did] = row["answer"]
        relevant_docs[qid] = {did}

    return queries, corpus, relevant_docs

def main():
    base_model_path = "/e/project1/reformo/enrico/models/jina-embeddings-v5-text-nano-retrieval"
    model_name = os.path.basename(os.path.normpath(base_model_path))
    seq_lenth = 512
    output_dir = f"model_output/{model_name}"
    final_output_dir = f"{output_dir}/final"
    test_set = "/e/project1/reformo/enrico/caselaw/test"
    eval_output=f"model_output/{model_name}/eval"
    resume=False
    matryoshka_dims = [768, 512, 256, 128]
    learning_rate=1e-6
    mini_batch_size=64
    eval_size=256

    # 1. Load a base model to finetune using FastSentenceTransformer
    model = SentenceTransformer(base_model_path)
    model.max_seq_length = seq_lenth

    # Load a dataset
    dataset = load_dataset(
        "parquet",
        data_dir="/e/project1/reformo/enrico/caselaw/train/",
        split="train",
        cache_dir="/e/project1/reformo/enrico/.hf_cache/datasets",
    )

    dataset_dict = dataset.train_test_split(test_size=10_000, seed=12)
    train_dataset: Dataset = dataset_dict["train"]
    eval_dataset: Dataset = dataset_dict["test"]

    print(train_dataset)
    print(eval_dataset)
    print(train_dataset[0])

    # 4. Build an Information Retrieval evaluator

    queries, corpus, relevant_docs = prepare_test(test_set)

    evaluator = InformationRetrievalEvaluator(
        queries=queries, 
        corpus=corpus, 
        relevant_docs=relevant_docs, 
        batch_size=eval_size,
        name=f"{model_name}-caselaw-test",
    )

    evaluator(model, output_path=eval_output)

    # 5. Define a loss function
    inner_train_loss = CachedMultipleNegativesRankingLoss(
        model, 
        mini_batch_size=mini_batch_size,
    )
    loss = MatryoshkaLoss(
        model, 
        inner_train_loss, 
        matryoshka_dims=matryoshka_dims
    )

    # 6. Define training args
    args = SentenceTransformerTrainingArguments(
        num_train_epochs=1,
        per_device_train_batch_size=2048,
        per_device_eval_batch_size=eval_size,
        learning_rate=learning_rate,
        logging_steps=10,
        optim="adamw_torch_fused",
        warmup_ratio=0.1,
        eval_strategy="steps",
        eval_steps=0.3,
        save_strategy="steps",
        save_steps=0.1,
        save_total_limit=3,
        fp16=False,
        bf16=True,
        output_dir=output_dir,
        lr_scheduler_type="cosine",
        batch_sampler=BatchSamplers.NO_DUPLICATES,
        run_name=f"{model_name}-caselaw",
    )

    # 7. Create a trainer & train the model
    trainer = SentenceTransformerTrainer(
        model=model,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        loss=loss,
        args=args,
        evaluator=evaluator,
    )

    # Track GPU memory usage for this LoRA run
    gpu_stats = torch.cuda.get_device_properties(0)
    start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
    max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
    print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
    print(f"{start_gpu_memory} GB of memory reserved.")

    trainer_stats = trainer.train(resume_from_checkpoint=resume)

    used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
    used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
    used_percentage = round(used_memory / max_memory * 100, 3)
    lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
    print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
    print(f"{round(trainer_stats.metrics['train_runtime'] / 60, 2)} minutes used for training.")
    print(f"Peak reserved memory = {used_memory} GB.")
    print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
    print(f"Peak reserved memory % of max memory = {used_percentage} %.")
    print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

    # (Optional) Evaluate the trained model on the evaluator after training
    evaluator(model, output_path=eval_output)

    model.save_pretrained(final_output_dir)

if __name__ == "__main__":
    main()
```
When I go to resume training loss goes back higher than it previously ended at and does not drop below.

Towards end of first run:
```
{'loss': '0.91', 'grad_norm': '0.8618', 'learning_rate': '3.094e-07', 'epoch': '0.6623'}
{'loss': '0.9154', 'grad_norm': '0.7672', 'learning_rate': '3.057e-07', 'epoch': '0.6645'}
{'loss': '0.9243', 'grad_norm': '0.7811', 'learning_rate': '3.021e-07', 'epoch': '0.6668'}
{'loss': '0.9205', 'grad_norm': '0.7763', 'learning_rate': '2.985e-07', 'epoch': '0.6691'}
{'loss': '0.9239', 'grad_norm': '0.7225', 'learning_rate': '2.948e-07', 'epoch': '0.6714'}
{'loss': '0.9009', 'grad_norm': '0.7645', 'learning_rate': '2.912e-07', 'epoch': '0.6736'}
{'loss': '0.9255', 'grad_norm': '0.8051', 'learning_rate': '2.877e-07', 'epoch': '0.6759'}
{'loss': '0.9113', 'grad_norm': '0.7445', 'learning_rate': '2.841e-07', 'epoch': '0.6782'}
{'loss': '0.9091', 'grad_norm': '0.7358', 'learning_rate': '2.805e-07', 'epoch': '0.6804'}
{'loss': '0.9233', 'grad_norm': '0.8034', 'learning_rate': '2.77e-07', 'epoch': '0.6827'}
{'loss': '0.9028', 'grad_norm': '0.8055', 'learning_rate': '2.734e-07', 'epoch': '0.685'}
{'loss': '0.9243', 'grad_norm': '0.8138', 'learning_rate': '2.699e-07', 'epoch': '0.6872'}
{'loss': '0.9028', 'grad_norm': '0.7221', 'learning_rate': '2.664e-07', 'epoch': '0.6895'}
{'loss': '0.9174', 'grad_norm': '0.7309', 'learning_rate': '2.629e-07', 'epoch': '0.6918'}
{'loss': '0.9168', 'grad_norm': '0.7746', 'learning_rate': '2.594e-07', 'epoch': '0.694'}
{'loss': '0.9244', 'grad_norm': '0.7616', 'learning_rate': '2.56e-07', 'epoch': '0.6963'}
{'loss': '0.9148', 'grad_norm': '0.9107', 'learning_rate': '2.525e-07', 'epoch': '0.6986'}
{'loss': '0.9092', 'grad_norm': '0.7964', 'learning_rate': '2.491e-07', 'epoch': '0.7008'}
{'loss': '0.9186', 'grad_norm': '0.7714', 'learning_rate': '2.457e-07', 'epoch': '0.7031'}
{'loss': '0.9132', 'grad_norm': '0.7656', 'learning_rate': '2.423e-07', 'epoch': '0.7054'}
{'loss': '0.9029', 'grad_norm': '0.7996', 'learning_rate': '2.389e-07', 'epoch': '0.7076'}
{'loss': '0.9042', 'grad_norm': '0.7516', 'learning_rate': '2.355e-07', 'epoch': '0.7099'}
```
When resuming:
```
{'loss': '1.196', 'grad_norm': '1.029', 'learning_rate': '2.491e-07', 'epoch': '0.7008'}
{'loss': '1.189', 'grad_norm': '0.9515', 'learning_rate': '2.457e-07', 'epoch': '0.7031'}
{'loss': '1.182', 'grad_norm': '0.9753', 'learning_rate': '2.423e-07', 'epoch': '0.7054'}
{'loss': '1.169', 'grad_norm': '1.032', 'learning_rate': '2.389e-07', 'epoch': '0.7076'}
{'loss': '1.171', 'grad_norm': '0.928', 'learning_rate': '2.355e-07', 'epoch': '0.7099'}
{'loss': '1.155', 'grad_norm': '0.9997', 'learning_rate': '2.322e-07', 'epoch': '0.7122'}
{'loss': '1.16', 'grad_norm': '1.024', 'learning_rate': '2.288e-07', 'epoch': '0.7144'}
{'loss': '1.175', 'grad_norm': '1.026', 'learning_rate': '2.255e-07', 'epoch': '0.7167'}
{'loss': '1.17', 'grad_norm': '1.006', 'learning_rate': '2.222e-07', 'epoch': '0.719'}
{'loss': '1.163', 'grad_norm': '0.9202', 'learning_rate': '2.189e-07', 'epoch': '0.7213'}
{'loss': '1.171', 'grad_norm': '0.8912', 'learning_rate': '2.157e-07', 'epoch': '0.7235'}
{'loss': '1.16', 'grad_norm': '1.022', 'learning_rate': '2.124e-07', 'epoch': '0.7258'}
{'loss': '1.172', 'grad_norm': '1.024', 'learning_rate': '2.092e-07', 'epoch': '0.7281'}
{'loss': '1.17', 'grad_norm': '1.015', 'learning_rate': '2.06e-07', 'epoch': '0.7303'}
{'loss': '1.154', 'grad_norm': '1.006', 'learning_rate': '2.028e-07', 'epoch': '0.7326'}
{'loss': '1.162', 'grad_norm': '0.9685', 'learning_rate': '1.996e-07', 'epoch': '0.7349'}
{'loss': '1.165', 'grad_norm': '1.093', 'learning_rate': '1.965e-07', 'epoch': '0.7371'}
{'loss': '1.163', 'grad_norm': '0.9454', 'learning_rate': '1.933e-07', 'epoch': '0.7394'}
{'loss': '1.179', 'grad_norm': '1.012', 'learning_rate': '1.902e-07', 'epoch': '0.7417'}
{'loss': '1.174', 'grad_norm': '0.9179', 'learning_rate': '1.871e-07', 'epoch': '0.7439'}
{'loss': '1.182', 'grad_norm': '1.019', 'learning_rate': '1.84e-07', 'epoch': '0.7462'}
{'loss': '1.158', 'grad_norm': '1.02', 'learning_rate': '1.81e-07', 'epoch': '0.7485'}
{'loss': '1.156', 'grad_norm': '0.9746', 'learning_rate': '1.779e-07', 'epoch': '0.7507'}
{'loss': '1.169', 'grad_norm': '0.947', 'learning_rate': '1.749e-07', 'epoch': '0.753'}
{'loss': '1.163', 'grad_norm': '0.9294', 'learning_rate': '1.719e-07', 'epoch': '0.7553'}
{'loss': '1.17', 'grad_norm': '1.038', 'learning_rate': '1.689e-07', 'epoch': '0.7575'}
{'loss': '1.174', 'grad_norm': '0.9617', 'learning_rate': '1.66e-07', 'epoch': '0.7598'}
{'loss': '1.165', 'grad_norm': '0.9768', 'learning_rate': '1.63e-07', 'epoch': '0.7621'}
{'loss': '1.174', 'grad_norm': '0.9736', 'learning_rate': '1.601e-07', 'epoch': '0.7643'}
{'loss': '1.155', 'grad_norm': '0.9279', 'learning_rate': '1.572e-07', 'epoch': '0.7666'}
{'loss': '1.17', 'grad_norm': '0.9773', 'learning_rate': '1.544e-07', 'epoch': '0.7689'}
{'loss': '1.175', 'grad_norm': '0.9271', 'learning_rate': '1.515e-07', 'epoch': '0.7711'}
{'loss': '1.157', 'grad_norm': '0.939', 'learning_rate': '1.487e-07', 'epoch': '0.7734'}
{'loss': '1.168', 'grad_norm': '0.9857', 'learning_rate': '1.459e-07', 'epoch': '0.7757'}
{'loss': '1.165', 'grad_norm': '0.9771', 'learning_rate': '1.431e-07', 'epoch': '0.778'}
{'loss': '1.152', 'grad_norm': '0.9378', 'learning_rate': '1.403e-07', 'epoch': '0.7802'}
{'loss': '1.173', 'grad_norm': '0.9425', 'learning_rate': '1.376e-07', 'epoch': '0.7825'}
{'loss': '1.153', 'grad_norm': '0.9958', 'learning_rate': '1.349e-07', 'epoch': '0.7848'}
{'loss': '1.155', 'grad_norm': '0.9647', 'learning_rate': '1.322e-07', 'epoch': '0.787'}
{'loss': '1.172', 'grad_norm': '1.019', 'learning_rate': '1.295e-07', 'epoch': '0.7893'}
{'loss': '1.16', 'grad_norm': '0.9941', 'learning_rate': '1.269e-07', 'epoch': '0.7916'}
{'loss': '1.179', 'grad_norm': '1.074', 'learning_rate': '1.243e-07', 'epoch': '0.7938'}
{'loss': '1.16', 'grad_norm': '0.9214', 'learning_rate': '1.217e-07', 'epoch': '0.7961'}
```
Any help would be appreciated.

Thank you,

Enrico

**Comment by conceptofmind** [comment]
Additionally, this warning is thrown:
```
[transformers] Warning: The following arguments do not match the ones in the `trainer_state.json` within the checkpoint directory: 
	eval_steps: 0.3 (from args) != 1323 (from trainer_state.json)
	save_steps: 0.1 (from args) != 441 (from trainer_state.json)
[transformers] Warning: The following arguments do not match the ones in the `trainer_state.json` within the checkpoint directory: 
	eval_steps: 0.3 (from args) != 1323 (from trainer_state.json)
	save_steps: 0.1 (from args) != 441 (from trainer_state.json)
[transformers] Warning: The following arguments do not match the ones in the `trainer_state.json` within the checkpoint directory: 
	eval_steps: 0.3 (from args) != 1323 (from trainer_state.json)
	save_steps: 0.1 (from args) != 441 (from trainer_state.json)
```

**Comment by ddofer** [comment]
I've had similar issues with resume/restart (even from checkpoints), and the issue was the cosine LR. i.e it restarted from a higher point. I don't know if that's the case here. (I didn't open up a bug about it to HF)
